executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,441 @@
+                                }
+                                MachineCallArgument::Known(e) => {
+                                    // convert known arguments into variable assignments that are then referenced
+                                    let idx = var_mapper.reserve_idx();
+                                    actions.push(InterpreterAction::AssignExpression(
+                                        idx,
+                                        var_mapper.map_expr_to_rpn(e),
+                                    ));
+                                    MachineCallArgumentIdx::Known(idx)
+                                }
+                            })
+                            .collect(),
+                    )
+                }
+                Effect::Branch(..) => {
+                    unimplemented!("Branches are not supported in the interpreter yet")
+                }
+            };
+            actions.push(action);
+        })
+    }
+
+    fn write_data(
+        var_mapper: &mut VariableMapper,
+        actions: &mut Vec<InterpreterAction<T>>,
+        effects: &[Effect<T, Variable>],
+    ) {
+        effects
+            .iter()
+            .flat_map(written_vars_in_effect)
+            .for_each(|(var, _mutable)| {
+                match var {
+                    Variable::Cell(cell) => {
+                        let idx = var_mapper.get_var(var).unwrap();
+                        actions.push(InterpreterAction::WriteCell(idx, cell.clone()));
+                    }
+                    Variable::Param(i) => {
+                        let idx = var_mapper.get_var(var).unwrap();
+                        actions.push(InterpreterAction::WriteParam(idx, *i));
+                    }
+                    Variable::MachineCallReturnValue(_) => {
+                        // This is just an internal variable.
+                    }
+                }
+            });
+    }
+
+    /// Check that actions are valid (e.g., variables written to only once, and only read after being written to)
+    fn is_valid(&self) -> bool {
+        let mut prev_writes = BTreeSet::new();
+        for action in &self.actions {
+            let writes = action.writes();
+            // writing to a variable already written?
+            if !writes.is_disjoint(&prev_writes) {
+                return false;
+            }
+            // reading a variable that was not written to?
+            if !action.reads().is_subset(&prev_writes) {
+                return false;
+            }
+            prev_writes.extend(writes);
+        }
+        true
+    }
+
+    /// Execute the machine effects for the given the parameters
+    pub fn call<Q: QueryCallback<T>>(
+        &self,
+        mutable_state: &MutableState<'_, T, Q>,
+        params: &mut [LookupCell<T>],
+        data: CompactDataRef<'_, T>,
+    ) {
+        let mut vars = vec![T::zero(); self.var_count];
```

Comment:
We could actually keep that one in the state - but can be done in another PR. There is also no need to reset it to zero between calls, since we checked that nothing is read before it is written.

Reply:
yep, i've done exactly this in the other branch I have